File: examples/scala-examples/src/main/scala/CustomDeserializer.scala
package com.gel.examples
import com.gel.driver.GelClientPool
import com.gel.driver.annotations.{GelDeserializer, GelName, GelType}
import org.slf4j.LoggerFactory

import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.FutureConverters.*

object CustomDeserializer:
  private val logger = LoggerFactory.getLogger(classOf[CustomDeserializer])

  @GelType
  class Person @GelDeserializer()
  (
    @GelName("name")
    val name: String,
    @GelName("age")
    val age: Long
  ) {
    logger.info("Custom deserializer called")
  }

class CustomDeserializer extends Example:
  import CustomDeserializer._

  override def run(clientPool: GelClientPool)(implicit context: ExecutionContext): Future[Unit] = {
    for(
      result <- client.queryRequiredSingle(
        classOf[Person],
        """
          | insert Person { name := 'Example', age := 123 } unless conflict on .name;
          | select Person { name, age } filter .name = 'Example'
          |""".stripMargin
      ).asScala
    ) yield {
      logger.info("Got person: name: {}, age: {}", result.name, result.age)
    }
  }

File: examples/java-examples/src/main/java/com/gel/examples/CustomDeserializer.java
package com.gel.examples;

import com.gel.driver.GelClientPool;
import com.gel.driver.annotations.GelDeserializer;
import com.gel.driver.annotations.GelName;
import com.gel.driver.annotations.GelType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletionStage;

public final class CustomDeserializer implements Example {
    private static final Logger logger = LoggerFactory.getLogger(CustomDeserializer.class);

    @GelType
    public static final class Person {
        private final String name;
        private final Long age;

        @GelDeserializer
        public Person(
                @GelName("name") String name,
                @GelName("age") Long age
        ) {
            this.name = name;
            this.age = age;

            logger.info("Custom deserializer called");
        }
    }

    @Override
    public CompletionStage<Void> run(GelClientPool clientPool) {
        return clientPool.execute("insert Person { name := 'Example', age := 123 } unless conflict on .name")
                .thenCompose(v -> clientPool.queryRequiredSingle(Person.class, "select Person { name, age } filter .name = 'Example'"))
                .thenAccept(person -> logger.info("Person result: {person}"));
    }
}